From: Michal Hocko <mhocko@suse.com>
Date: Wed, 14 Jun 2017 08:18:00 +0200
Subject: mm: do not collapse stack gap into THP
Bug-Debian: https://security-tracker.debian.org/tracker/CVE-2017-1000364

Oleg has noticed that khugepaged will happilly collapse stack vma (as
long as it is not an early stack - see is_vma_temporary_stack) and
it might effectively remove the stack gap area as well because a larger
part of the stack vma is usually populated. The same applies to the
page fault handler.

Fix this by checking stack_guard_area when revalidating a VMA
in hugepage_vma_revalidate.  We do not want to hook/replace
is_vma_temporary_stack() check because THP might be still useful for
stack, all we need is excluding the gap from collapsing into a THP.

Also check the to-be-created THP in do_huge_pmd_anonymous_page to
make sure it is completely outside of the gap area because we we could
create THP covering the gap area.

Noticed-by: Oleg Nesterov <oleg@redhat.com>
Signed-off-by: Michal Hocko <mhocko@suse.com>
[carnil: backport to 3.16: move stack_guard_area check to huge_memory.c
in collapse_huge_page]
[bwh: Backported to 3.2:
 - Use 'goto' rather than 'return VM_FAULT_FALLBACK' to fall back to PTEs
 - Adjust context]
---
 mm/huge_memory.c | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/mm/huge_memory.c b/mm/huge_memory.c
index d6e6cafdb2c9..ef201fe46804 100644
--- a/mm/huge_memory.c
+++ b/mm/huge_memory.c
@@ -680,6 +680,9 @@ int do_huge_pmd_anonymous_page(struct mm_struct *mm, struct vm_area_struct *vma,
 	pte_t *pte;
 
 	if (haddr >= vma->vm_start && haddr + HPAGE_PMD_SIZE <= vma->vm_end) {
+		if (stack_guard_area(vma, haddr) ||
+				stack_guard_area(vma, haddr + HPAGE_PMD_SIZE))
+			goto out;
 		if (unlikely(anon_vma_prepare(vma)))
 			return VM_FAULT_OOM;
 		if (unlikely(khugepaged_enter(vma, vma->vm_flags)))
@@ -1922,6 +1925,10 @@ static void collapse_huge_page(struct mm_struct *mm,
 		goto out;
 	if (!hugepage_vma_check(vma))
 		goto out;
+
+	/* never try to collapse stack gap */
+	if (stack_guard_area(vma, hstart) || stack_guard_area(vma, hend))
+		goto out;
 	pgd = pgd_offset(mm, address);
 	if (!pgd_present(*pgd))
 		goto out;
